### PR TITLE
Rewrites cpu cnn implementation for compatibility with dl4j implementation

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?

Rewrites c++ conv2d implementation to be inline with the dl4j implementation.
Fixes gradient check failures.

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

**Comment by treo** [comment]
This looks like it includes #10081? Can you update the branch?

**Comment by agibsonccc** [comment]
> This looks like it includes #10081? Can you update the branch?

Willl do

**Comment by agibsonccc** [comment]
@treo it looks like it's down to 6 files now.

File: libnd4j/include/ops/declarable/helpers/cpu/im2col.cpp [hunk 1]
```diff
@@ -61,87 +73,52 @@ static void im2col_(sd::LaunchContext& context, const NDArray& input, NDArray& o
...
         }
       }
-    };
+    }
+  };
 
-    samediff::Threads::parallel_for(func, 0, bS, 1, 0, oH, 1);
-  }
+  samediff::Threads::parallel_for(func, 0, bS, 1, 0, oH, 1);
+  output.printBufferRaw("IM2COL OUTPUT BUFFER:");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cpu/im2col.cpp]: This looks like a stray debug output?

File: libnd4j/include/ops/declarable/helpers/cpu/convolutions_conv2d.cpp [hunk 2]
```diff
@@ -23,89 +23,97 @@
...
+
   //----- calculation of output -----//
   auto ctx = block.launchContext();
-  helpers::im2col(
-      *ctx, *input, colP, kH, kW, sH, sW, pH, pW, dH, dW,
-      NDArrayFactory::create(0.f, input->getContext()));  // [bS, iC, iH, iW] is convoluted to [bS, iC, kH, kW, oH, oW]
-  MmulHelper::tensorDot(&col, weights, &mmulResult, {3, 4, 5}, wAxes,
-                        {});  // [bS, oH, oW, kH, kW, iC] x [kH, kW, iC, oC] = [bS, oH, oW, oC]
+  input->printIndexedBuffer("im2col INPUT BUFFER:");
+  input->printShapeInfo("im2col INPUT SHAPE INFO:");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cpu/convolutions_conv2d.cpp]: Stray debug print?

File: libnd4j/include/ops/declarable/helpers/cpu/convolutions_conv2d.cpp [hunk 3]
```diff
@@ -23,89 +23,97 @@
...
+                    NDArrayFactory::create(0.f, input->getContext()));
+  } else {
+    // For NHWC, we need to permute the input to NCHW before im2col
+    NDArray* inputNchw = new NDArray(input->permute({0, 3, 1, 2}));
+    helpers::im2col(*ctx, *inputNchw, *colP, kH, kW, sH, sW, pH, pW, dH, dW,
+                    NDArrayFactory::create(0.f, input->getContext()));
+  }
+
+  col->printBufferRaw("col post permute forward pass: ");
+  colP->printBufferRaw("colP post permute forward pass: ");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cpu/convolutions_conv2d.cpp]: Stray debug print?

File: libnd4j/include/ops/declarable/helpers/cpu/convolutions_conv2d.cpp [hunk 4]
```diff
@@ -23,89 +23,97 @@
...
+
+  std::vector<sd::LongType> permute = {0, 3, 4, 5, 1, 2};
+  block.pushIntermediateResult(col);
+
+  auto im2colReshape = col->reshape('c', {bS * oH * oW, kH * kW * iC}, true);
+  im2colReshape.printBufferRaw("im2col post reshape forward pass: ");
+  im2colReshape.printIndexedBuffer("IM2col after reshape:");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cpu/convolutions_conv2d.cpp]: Stray debug print?

File: libnd4j/include/ops/declarable/helpers/cpu/im2col.cpp [hunk 5]
```diff
@@ -28,10 +28,22 @@ namespace helpers {
...
+    THROW_EXCEPTION("ops::helpers::im2col: input array must have rank = 4");
+  }
+
+  if(output.rankOf() != 6) {
+    THROW_EXCEPTION("ops::helpers::im2col: output array must have rank = 6");
+  }
+
+  input.printBufferRaw("INPUT IM2COL BUFFER:");
+
+  input.printIndexedBuffer("IM2COL INPUT:");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cpu/im2col.cpp]: Stray debug print?

File: libnd4j/include/ops/declarable/generic/nn/convo/conv1d.cpp [hunk 6]
```diff
@@ -227,86 +211,134 @@ CUSTOM_OP_IMPL(conv1d_bp, 3, 2, false, 0, 5) {
...
+      gradB->assign(gradB);
+    }
+  }
 
-  sd::ops::conv2d_bp conv2dBP;
-  auto status = conv2dBP.execute({&inputReshaped, &weightsReshaped, bias, &gradOReshaped},
-                                 {&gradIReshaped, &gradWReshaped, gradB}, {},
-                                 {1, kW, 1, sW, 0, pW, 1, dW, paddingMode, !isNCW, wFormat}, {});
-  if (status != sd::Status::OK) return status;
+  gradW->printIndexedBuffer("GRAD W RESHAPED AFTER:");
```
Comment (treo) [inline libnd4j/include/ops/declarable/generic/nn/convo/conv1d.cpp]: Stray debug print?

File: libnd4j/include/ops/declarable/generic/nn/convo/conv1d.cpp [hunk 7]
```diff
@@ -227,86 +211,134 @@ CUSTOM_OP_IMPL(conv1d_bp, 3, 2, false, 0, 5) {
...
+                             {1, kW, 1, sW, 0, pW, 1, dW, paddingMode, !isNCW, wFormat}, {});
+    }
+
+  }
+
+  if(gradIReshaped->buffer() != gradI->buffer()) {
+    gradI->assign(gradIReshaped);
+  }
+
+  gradW->printIndexedBuffer("GRAD W RESHAPED BEFORE:");
```
Comment (treo) [inline libnd4j/include/ops/declarable/generic/nn/convo/conv1d.cpp]: Stray debug print?

File: libnd4j/include/ops/declarable/generic/nn/convo/conv1d.cpp [hunk 8]
```diff
@@ -189,29 +175,27 @@ CUSTOM_OP_IMPL(conv1d_bp, 3, 2, false, 0, 5) {
   auto gradI = OUTPUT_NULLIFIED(0);                                // [bS, iW, iC] (NWC) or [bS, iC, iW] (NCW), epsilon
   auto gradW = OUTPUT_NULLIFIED(1);                                // [kW, iC, oC], [oC, iC, kW], [oC, kW, iC]
   auto gradB = block.width() > 3 ? OUTPUT_NULLIFIED(2) : nullptr;  // [oC]
-
+  gradI->printIndexedBuffer("gradI");
+  gradW->printIndexedBuffer("gradW");
+  gradB->printIndexedBuffer("gradB");
```
Comment (treo) [inline libnd4j/include/ops/declarable/generic/nn/convo/conv1d.cpp]: Stray debug print?

**Review by treo — CHANGES_REQUESTED** [review]
Looks like there are a couple of stray debug prints left over.

**Comment by treo** [comment]
I can still find plenty of left-over prints, even among those that I've marked. Check the changes in this PR here on github and do a search for print after loading all of them, that should find all of them.

**Comment by agibsonccc** [comment]
Ah I missed conv1d. I did a fresh scan with printIndexedBuffer/printShapeInfo. That should be it.

**Review by treo — APPROVED** [review]

**Pull request merged**